src/hostapi/alsa/pa_linux_alsa.c

```diff
@@ -3156,6 +3160,28 @@ static PaError IsStreamActive( PaStream *s )
     return stream->isActive;
 }
 
+/** Extract audio/trigger htstamp from status and convert into PaTime (seconds).
+ *
+ * trigger is boolean:  trigger stampstamp vs audio timestamp.  If delay is non-NULL, return delay in
+ * frames.  */
+static PaTime StatusToTime( const snd_pcm_status_t *status, int trigger, snd_pcm_uframes_t* delay )
+{
+    snd_htimestamp_t timestamp;
+    if ( trigger )
+    {
+        alsa_snd_pcm_status_get_trigger_htstamp( status, &timestamp );
+    }
+    else
+    {
+        alsa_snd_pcm_status_get_htstamp( status, &timestamp );
+    }
+    if ( delay )
+    {
+        *delay = alsa_snd_pcm_status_get_delay( status );
+    }
+    return timestamp.tv_sec + ( (PaTime)timestamp.tv_nsec / 1000000000.0 );
```

Comment:
Better to multiply my reciprocal here.

Reply:
Also please avoid long strings of zeros. Prefer:

```suggestion
    return timestamp.tv_sec + ( (PaTime)timestamp.tv_nsec * 1.0E-9 );
```

Reply:
Ok, changed.  It's not exactly the same math, since 1e-9 isn't an exact floating point number, while 1e9 is.  But I don't think any possible values of tv_nsec result in a different result in the end.

Reply:
You raise a valid point about division potentially preserving more accuracy.  Using reciprocal is desirable for performance reasons, but I would prefer not to throw away usable accuracy.  'PaStreamCallbackTimeInfo' should retain all available accuracy within the limits of what is representable by PaTime (64-bit float).

Reply:
I ran a test that measured the difference between (n/1e9) and (n*1e-9) for all possible n.
The maximum error was  1.11022e-16 when n =  500000001
That seems like an insignificant numeric error. I think the performance hit is more significant.

```
void findMaxDelta() {
    double maxDelta = 0.0;
    uint xAtMax = -1;
    for (uint x = 1; x < 1e9; x++) {
        double y1 = x / 1.0e9;
        double y2 = x * 1.0e-9;
        double yd = fabs(y1 - y2);
        if (yd > maxDelta) {
            maxDelta = yd;
            xAtMax = x;
            std::cout << x << ", " << y1 << ", " << y2 << ", " << yd << "\n";
        }
    }
    std::cout << "xAtMax  = " << xAtMax << ", maxDelta = " << maxDelta << "\n";
}
```

Reply:
And since the timestamp is only precise to the ns to begin with, once you round that result with max error to nearest nanosecond, it becomes the exact same value as it would have had with no error.  That's what I meant when I said, "I don't think any possible values of tv_nsec result in a different result in the end."

Reply:
1.11022e-16 seems insignificant to me. It is a small fraction of a nanosecond, and float64 only maintains about 13 decimal digits of precision, which by the time `timestamp.tv_sec` is added will cause quite a few digits to be lost. There are 31536000 seconds in a year so after a year the final result of `timestamp.tv_sec + ( (PaTime)timestamp.tv_nsec * 1e-9 )` will only include about 6-7 significant digits of the nanosecond part.